File: Go/150.go
/*
iterate thru the array, push to stack if the string is number.
If the string is operator, pop top 2 numbers from the stack and push the computed result to stack.
return top element on the stack.
*/
func evalRPN(tokens []string) int {
    leng, top := len(tokens), -1
    if leng == 0 {
        return 0
    }
    stack := make([]int, leng, leng)
    for _, s := range tokens {
        switch(s){
        case "+":
            n1, n2 := stack[top-1], stack[top]
            top--
            stack[top] = n1 + n2
        case "-":
            n1, n2 := stack[top-1], stack[top]
            top--
            stack[top] = n1 - n2
        case "*":
            n1, n2 := stack[top-1], stack[top]
            top--
            stack[top] = n1 * n2
        case "/":
            n1, n2 := stack[top-1], stack[top]
            top--
            stack[top] = n1 / n2
        default:
            if num, err := strconv.Atoi(s); err == nil {
                top++
                stack[top] = num
            }
        }
    }
    return stack[top]
}

File: Java/150.java
/*
put numbers into the stack, and once we encounter an operator, 
pop 2 numbers and calculate, push the result back to stack.
*/
class Solution {
    public int evalRPN(String[] tokens) {
        if(tokens == null || tokens.length == 0) return 0;
        
        int[] stack = new int[tokens.length];
        int top = 0;
        
        for(String token : tokens){
            switch(token){
                case "+":
                    int sum = stack[top-1] + stack[top-2];
                    top -= 2;
                    stack[top++] = sum;
                    break;
                case "-":
                    int sub = stack[top-2] - stack[top-1];
                    top -= 2;
                    stack[top++] = sub;
                    break;
                case "*":
                    int mul = stack[top-1] * stack[top-2];
                    top -= 2;
                    stack[top++] = mul;
                    break;
                case "/":
                    int div = stack[top-2] / stack[top-1];
                    top -= 2;
                    stack[top++] = div;
                    break;
                default:
                    stack[top++] = Integer.valueOf(token);
            }
        }
        return stack[0];
    }
}
